henryPuma/assignments/treesBFS/lc102/lc102.js

```diff
@@ -0,0 +1,130 @@
+// Problem Statement #
+
+// Given a binary tree, populate an array to represent its level-by-level traversal. You should populate the values of all nodes of each level from left to right in separate sub-arrays.
+
+class TreeNode {
+
+  constructor(value) {
+    this.value = value;
+    this.left = null;
+    this.right = null;
+  }
+};
+
+// Precondition:
+// root: The head of the binary tree
+// Postcondition:
+// levels: an array containing all the levels of the input binary tree that is referenced by root
+
+// global: An array (queue) that will process the current nodes children
+// currentLevel: A queue that temporarily holds the current level of the tree
+// levels: the returning containing all the levels of the binary treeChildren
+
+// currentNode: The node that gets popped from the queue to be processed
+
+// globalLength: stores the length of the global arrays
+
+/*General Idea:
+ Use a 'levels' array to store the final solution,
+ Use a 'global' array to store nodes that are in the next level and need processing,
+ Use 'globalLength' variable to keep track of the number of nodes that are being processed,
+ Use a 'currentLevel' to temporarily keep track of all processed nodes in the current level, 
+ While loop: We process all nodes in the global array, hence the reason we check for global.length
+ For loop: Process current level nodes (push to temp currentLevel array) AND add next level nodes to be processed for the next iteration,
+ After the for loop, we add all the processed nodes in the 'currentLevel' array to the 'levels' array,
+ Rinse and repeat until there are no longer nodes in the 'global' array that needs processing
+ */
+
+function traverse(root) {
+  // root is null
+  if (!root) return [];
+  
+  const levels = []; // Solution array
+  const global = [root]; // Contains nodes that needs processing
+
+  let currentLevel = [];
+  let globalLength = global.length;
+
+  // While there are nodes that need to be processed
+  while(global.length){
+
+    globalLength = global.length; // Update the amount of nodes that need processing
+
+    for (i = 0; i < globalLength; i++){
+      currentNode = global.shift();
+      if(currentNode.left) global.push(currentNode.left);
+      if(currentNode.right) global.push(currentNode.right);
+      currentLevel.push(currentNode.value); // Store processed node for current level
+    }
+    levels.push(currentLevel); // Push all processed nodes
+    currentLevel = [];
+  }
+    
+  // Runtime: O(n)
+  // Space: O(n)
```

Comment:
space and time can actually be added to your high level overview, sometimes interviewer might like to see a small explanation as to why these are the way they are